File: packages/types/src/thread/variables.ts
/**
 * Thread Variable Type Definitions
 */

import type { Metadata } from "../common.js";
import type { VariableScope } from "./scopes.js";

/**
 * Variable Value Type
 */
export type VariableValueType = "number" | "string" | "boolean" | "array" | "object";

/**
 * Thread Variable Types
 */
export interface ThreadVariable {
  /** variable name */
  name: string;
  /** variable value */
  value: unknown;
  /** Variable type */
  type: VariableValueType;
  /** variable scope */
  scope: VariableScope;
  /** Read-only or not */
  readonly: boolean;
  /** variable metadata */
  metadata?: Metadata;
}

File: crates/wf-types/src/workflow_execution/variables.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum VariableValueType {
    Number,
    String,
    Boolean,
    Array,
    Object,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct VariableDefinition {
    pub name: String,
    pub value: serde_json::Value,
    pub r#type: Option<VariableValueType>,
    pub scope: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub readonly: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<crate::Metadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct VariableScope {
    pub variables: HashMap<String, serde_json::Value>,
    pub parent_scope: Option<Box<VariableScope>>,
}
